test/src/test/java/hudson/cli/ReloadConfigurationCommandTest.java

```diff
@@ -0,0 +1,206 @@
+        replace(node.getNodeName().equals("") ? "config.xml" : String.format("nodes/%s/config.xml",node.getNodeName()), "oldLabel", "newLabel");
+
+        assertThat(node.getLabelString(), equalTo("oldLabel"));
+
+        reloadJenkinsConfigurationViaCliAndWait();
+    }
+
+    @Test
+    public void reloadUserConfig() throws Exception {
+        User user = User.get("some_user", true, null);
+        user.setFullName("oldName");
+        user.save();
+
+        replace("users/some_user/config.xml", "oldName", "newName");
+
+        assertThat(user.getFullName(), equalTo("oldName"));
+
+        reloadJenkinsConfigurationViaCliAndWait();
+
+        assertThat(user.getFullName(), equalTo("newName"));
+    }
+
+    @Test
+    public void reloadJobConfig() throws Exception {
+        FreeStyleProject project = j.createFreeStyleProject("a_project");
+        project.setDescription("oldDescription");
+
+        replace("jobs/a_project/config.xml", "oldDescription", "newDescription");
+
+        assertThat( project.getDescription(), equalTo("oldDescription"));
+
+        reloadJenkinsConfigurationViaCliAndWait();
+
+        project = j.jenkins.getItem("a_project", j.jenkins, FreeStyleProject.class);
+        assertThat(project.getDescription(), equalTo("newDescription"));
+    }
+
+    @Test
+    public void reloadViewConfig() throws Exception {
+        ListView view = new ListView("a_view");
+        j.jenkins.addView(view);
+
+        view.setIncludeRegex("oldIncludeRegex");
+        view.save();
+
+        replace("config.xml", "oldIncludeRegex", "newIncludeRegex");
+
+        assertThat(view.getIncludeRegex(), equalTo("oldIncludeRegex"));
+
+        reloadJenkinsConfigurationViaCliAndWait();
+
+        view = (ListView) j.jenkins.getView("a_view");
+        assertThat(view.getIncludeRegex(), equalTo("newIncludeRegex"));
+    }
+
+    @Ignore // Until fixed JENKINS-8217
+    @Test
+    public void reloadDescriptorConfig() throws Exception {
+        Mailer.DescriptorImpl desc = j.jenkins.getExtensionList(Mailer.DescriptorImpl.class).get(0);;
+        desc.setDefaultSuffix("@oldSuffix");
+        desc.save();
+
+        replace("hudson.tasks.Mailer.xml", "@oldSuffix", "@newSuffix");
+
+        assertThat(desc.getDefaultSuffix(), equalTo("@oldSuffix"));
+
+        reloadJenkinsConfigurationViaCliAndWait();
+
+        assertThat(desc.getDefaultSuffix(), equalTo("@newSuffix"));
+    }
+
+    private void reloadJenkinsConfigurationViaCliAndWait() throws Exception {
+        final CLICommandInvoker.Result result = command
+                .authorizedTo(Jenkins.READ, Jenkins.ADMINISTER).invoke();
+
+        assertThat(result, succeededSilently());
+
+        // reload-configuration is performed in a separate thread
+        // we have to wait until it finishes
+        boolean sleep = true;
+
+        while (sleep) {
+            sleep = false;
+            for (Thread t : Thread.getAllStackTraces().keySet()) {
```

Comment:
Waiting for `j.jenkins.servletContext.getAttribute("app") instanceof Jenkins` is cleaner.